### PR TITLE
add raw socket (AF_PACKET) backend for virtio-net host

### PR DESCRIPTION
**Pull request by thehajime**
this includes two commits: 1) refactor the tap code to support generic file descriptor handling, 2) add raw socket backend for virtio-net.

---

This change is [Reviewable](https://reviewable.io/reviews/lkl/linux/171)

### CODE REVIEW

**Comment by opurdila** [comment]
Nice, thanks Hajime ! LGTM, I'll wait a couple of days for more reviews before merging.

File: tools/lkl/lib/virtio_net_linux_fdnet.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+extern struct lkl_netdev_linux_fdnet_ops {
+	/*
+	 * We need this so that we can "unhijack" this function in
+	 * case we decided to hijack it.
+	 */
+	int (*eventfd)(unsigned int initval, int flags);
+} lkl_netdev_linux_fdnet_ops;
+
+struct lkl_netdev_linux_fdnet *lkl_register_netdev_linux_fdnet(int fd);
+void lkl_unregister_netdev_linux_fdnet(struct lkl_netdev_linux_fdnet *nd);
```
Comment (liuyuan10) [inline tools/lkl/lib/virtio_net_linux_fdnet.h]: is lkl_unregister_netdev_linux_fdnet used anywhere externally?

**Comment by liuyuan10** [comment]
Do you have performance numbers vs tap? If it bypass kernel efficiently, it should give better numbers.

File: tools/lkl/lib/virtio_net_linux_fdnet.h (hunk 1 above)
Comment (thehajime) [inline tools/lkl/lib/virtio_net_linux_fdnet.h]: no at this moment, but I intended this to be used externally.

(I added comments for those function prototypes)

**Comment by thehajime** [comment]
> Do you have performance numbers vs tap? If it bypass kernel efficiently, it should give better numbers.

this is numbers btw/ raw socket and tap (also put in the commit message).

those are based on hijacked netperf  (client) to native netserver (server) connected via 10Gbps back-to-back link.
the raw socket result uses outgoing NIC directly while the tap involves host bridge (br0 consists of tap and the outgoing NIC) so, I guess it's not apple-to-apple comparison.

```
- TCP_RR
 raw(QDISK_BYPASS): 9519.31 Trans/sec
 tap:               9486.03 Trans/sec
- TCP_STREAM
 raw(QDISK_BYPASS): 2184.79 Mbps
 tap:               2130.39 Mbps
- UDP_STREAM
 raw(QDISK_BYPASS): 3654.32 Mbps
 tap:               3108.10 Mbps
```

the improvements are not that significant, and I guess without QDISK_BYPASS (i.e., the host kernel is before Linux 3.14), then the differences might be trivial too.

File: tools/lkl/lib/virtio_net_linux_fdnet.c [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+
+struct lkl_netdev_linux_fdnet *lkl_register_netdev_linux_fdnet(int fd)
+{
+	struct lkl_netdev_linux_fdnet *nd;
+
+	nd = (struct lkl_netdev_linux_fdnet *)
+		malloc(sizeof(struct lkl_netdev_linux_fdnet));
+	if (!nd) {
+		fprintf(stderr, "raw: failed to allocate memory\n");
```
Comment (liuyuan10) [inline tools/lkl/lib/virtio_net_linux_fdnet.c]: "raw" in error msg?

File: tools/lkl/lib/virtio_net_linux_fdnet.c [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+		return NULL;
+	}
+
+	if (lkl_netdev_linux_fdnet_ops.eventfd) {
+		/* eventfd is supported by the host, all is well */
+		nd->eventfd = lkl_netdev_linux_fdnet_ops.eventfd(
+			0, EFD_NONBLOCK | EFD_SEMAPHORE);
+
+		if (nd->eventfd < 0) {
+			perror("lkl_netdev_raw_create eventfd");
```
Comment (liuyuan10) [inline tools/lkl/lib/virtio_net_linux_fdnet.c]: "raw" in error msg?

**Comment by liuyuan10** [comment]
LGTM
Thanks for the detailed numbers

File: tools/lkl/lib/virtio_net_linux_fdnet.c (hunk 2 above)
Comment (thehajime) [inline tools/lkl/lib/virtio_net_linux_fdnet.c]: fixed. thanks.

File: tools/lkl/lib/virtio_net_linux_fdnet.c (hunk 3 above)
Comment (thehajime) [inline tools/lkl/lib/virtio_net_linux_fdnet.c]: fixed too.

**Comment by thehajime** [comment]
updated the patch set with the comment @liuyuan10.

**Pull request merged**